### PR TITLE
Refactor REST API v1

### PR DESCRIPTION
**Pull request by holmboe**
- Use the REST API
- Output nicely formatted tables
- Support both network scans and web scans

### CODE REVIEW

File: holmscan/cli.py [hunk 1]
```diff
@@ -77,10 +175,34 @@ def parse_cli():
...
+        if sub_args["asset"]:
+            sub_args = docopt(net_asset_args, argv=argv)
+        elif sub_args["profile"]:
+            sub_args = docopt(net_profile_args, argv=argv)
+        elif sub_args["scan"]:
+            sub_args = docopt(net_scan_args, argv=argv)
+        #elif sub_args["schedule"]:
+        #    sub_args = docopt(net_schedule_args, argv=argv)
+        else:
+            extras(
```
Comment (Grokzen) [inline holmscan/cli.py]: This extras call seems to be present multiple times, maybe abstract into a separate call, or look at maybe having it as a globally "else" case that it defaults out to. I am thinking like
```
def print_help(base_args):
  extras(True, holmsan.__version__, [Option('-h', '--help', 0, True)], base_args
```

File: holmscan/cli.py [hunk 2]
```diff
@@ -97,24 +219,102 @@ def run(cli_args, sub_args):
...
-            print(c.scan.get_web_scan(scan_id))
+        if cli_args["<command>"] == "net" and sub_args.get("asset", False):
+            data = c.scan.get_net_assets()
+            log.debug(pformat(data))
+            filtered = [[x["name"], x["uuid"]] for x in data["results"]]
+            print(tabulate(filtered, headers=["Name", "UUID"], **tabulate_args))
+        elif cli_args["<command>"] == "net" and sub_args.get("profile", False):
+            data = c.scan.get_net_profiles()
+            log.debug(pformat(data))
+            filtered = [[x["name"], x["uuid"]] for x in data]
```
Comment (Grokzen) [inline holmscan/cli.py]: Single names variables should be avoided if possible. Add a meaning to the variable or default to the more common `item` as var name.

File: holmscan/cli.py [hunk 3]
```diff
@@ -97,24 +219,102 @@ def run(cli_args, sub_args):
...
-                node_source_overrides=False,
-                is_was_discovery=True,
-                assets=[asset],
-                scheduled=False,
-            )
-            print(c.scan.get_web_scan(scan_id))
+        if cli_args["<command>"] == "net" and sub_args.get("asset", False):
+            data = c.scan.get_net_assets()
+            log.debug(pformat(data))
+            filtered = [[x["name"], x["uuid"]] for x in data["results"]]
```
Comment (Grokzen) [inline holmscan/cli.py]: Same here `x` -> `item`

File: holmscan/cli.py [hunk 4]
```diff
@@ -97,24 +219,102 @@ def run(cli_args, sub_args):
...
 
+                filtered = [
+                    [
+                        x["started_date"],
+                        x["finished_date"],
+                        x["status"],
+                        x["vulnerabilities_count"],
+                        x["uuid"],
+                    ]
+                    for x in data["results"]
```
Comment (Grokzen) [inline holmscan/cli.py]: `x` -> `item`

File: holmscan/constants.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
+
+SCAN_STATUS_CHOICES = ["completed", "running"]
```
Comment (Grokzen) [inline holmscan/constants.py:3]: When doing choices constants, even the choices should be made into constants.
```
SCAN_STATUS_COMPLETED = 'completed'
SCAN_STATUS_RUNNING = 'running'

SCAN_STATUS_CHOICES = [
    SCAN_STATUS_COMPLETED,
    SCAN_STATUS_RUNNING,
]
```

File: holmscan/controller.py [hunk 6]
```diff
@@ -22,24 +23,18 @@
 logging.getLogger("anyconfig").setLevel(logging.ERROR)
 
 
-CONFIGURABLES = [
-    "HOLMSCAN_DEBUG",
-    "HOLMSEC_PASSWORD",
-    "HOLMSEC_URL",
-    "HOLMSEC_USERNAME",
-]
+CONFIGURABLES = ["HOLMSCAN_DEBUG", "HOLMSEC_ENDPOINT", "HOLMSEC_TOKEN"]
```
Comment (Grokzen) [inline holmscan/controller.py:26]: Some of these are technically constants and maybe should be moved to constants.py file. But some of them fits more to have in here anyway.

File: holmscan/controller.py [hunk 7]
```diff
@@ -22,24 +23,18 @@
...
 DEFAULTS = {
     "HOLMSCAN_DEBUG": False,
-    "HOLMSEC_PASSWORD": "",
-    "HOLMSEC_URL": "https://sc.holmsecurity.com/",
-    "HOLMSEC_USERNAME": "",
+    "HOLMSEC_ENDPOINT": "https://se-api.holmsecurity.com/v1",
```
Comment (Grokzen) [inline holmscan/controller.py:29]: The variable name `DEFAULTS` is to non descriptive. It do not tell what kind or defaults for what purpose.

File: holmscan/controller.py [hunk 8]
```diff
@@ -22,24 +23,18 @@
...
+    "HOLMSEC_TOKEN": "",
 }
-REQUIRED = ["HOLMSEC_PASSWORD", "HOLMSEC_URL", "HOLMSEC_USERNAME"]
-VALIDATORS = {"HOLMSEC_URL": "^http(s)?://[a-zA-Z0-9._-]+/$"}
-VALID_EXAMPLES = {"HOLMSEC_URL": "example: https://sc.holmsecurity.com/"}
+REQUIRED = ["HOLMSEC_ENDPOINT", "HOLMSEC_TOKEN"]
```
Comment (Grokzen) [inline holmscan/controller.py:32]: `REQUIRED` is to non descriptive. Required for what?

File: holmscan/controller.py [hunk 9]
```diff
@@ -22,24 +23,18 @@
...
+VALIDATORS = {"HOLMSEC_ENDPOINT": "^http(s)?://[a-zA-Z0-9._-]+/v[0-9]$"}
+VALID_EXAMPLES = {"HOLMSEC_ENDPOINT": "example: https://se-api.holmsecurity.com/v1"}
 CONFIG_EXAMPLES = {
-    "HOLMSEC_USERNAME": "example: export HOLMSEC_USERNAME=username@example.com",
-    "HOLMSEC_URL": "example: echo HOLMSEC_URL: https://sc.holmsecurity.com/ >> ~/.config/holmscan.yaml",
+    "HOLMSEC_ENDPOINT": "example: echo HOLMSEC_ENDPOINT: https://se-api.holmsecurity.com/v1 >> ~/.config/holmscan.yaml",
```
Comment (Grokzen) [inline holmscan/controller.py:36]: `"HOLMSEC_ENDPOINT"` looks like a constant. Move it to
```
HOLMSEC_ENDPOINT = "HOLMSEC_ENDPOINT"
```

File: holmscan/scan.py [hunk 10]
```diff
@@ -16,73 +15,80 @@ class Scan(HolmscanModule):
...
-        )
-
-        job_url = u"{0}scan/wscans/".format(self.controller.conf.get("HOLMSEC_URL"))
-        job_response = self.controller.session.post(
-            job_url,
-            data=json.dumps(data).replace(" ", ""),
-            headers={"Content-Type": "application/json"},
-        )
+    def get_net_assets(self):
+        url = "{}/net-scans/assets".format(self.controller.conf.get("HOLMSEC_ENDPOINT"))
```
Comment (Grokzen) [inline holmscan/scan.py]: All `{}` formats should use positional arguments saying what item it should use. `{0}, {1} etc`

File: holmscan/scan.py [hunk 11]
```diff
@@ -16,73 +15,80 @@ class Scan(HolmscanModule):
...
+        headers = {
+            "Authorization": "Token {}".format(
+                self.controller.conf.get("HOLMSEC_TOKEN")
+            )
+        }
+        response = self.controller.session.get(url, headers=headers)
```
Comment (Grokzen) [inline holmscan/scan.py]: Having to build up the authorization header for every call outside the real call is redundant and prone to errors. Move the headers build into the `.get()` call. If the .get call is the requests session object, then consider to make a wrapper call in the self.controller class to expose a .get() call that constructs the header and just passes it along to the session.get() method call. To preserve the feature of having the ability to add more header fields, a kwargs could be accepted in the new .get() method that would merge with the authorization header into a single dict and then passed to requests session.

File: holmscan/scan.py [hunk 12]
```diff
@@ -16,73 +15,80 @@ class Scan(HolmscanModule):
...
-        assets = json.loads(self.controller.session.get(assets_url).text)
-        for asset in assets["results"]:
-            # TODO: is case insensitive a good thing for asset names?
-            if name.lower() == asset["name"].lower():
-                return asset["id"]
-
-        webapps_url = u"{0}scan/webapps/".format(
-            self.controller.conf.get("HOLMSEC_URL")
+        log.debug("Sending to {}".format(url))
+        headers = {
```
Comment (Grokzen) [inline holmscan/scan.py]: Same headers thingy here.

File: setup.py [hunk 13]
```diff
@@ -17,7 +17,7 @@
 with open("CHANGELOG.md") as f:
     CHANGELOG = f.read()
 
-install_requires = ["anyconfig", "appdirs", "docopt", "pyyaml", "requests"]
+install_requires = ["anyconfig", "appdirs", "docopt", "pyyaml", "requests", "tabulate"]
```
Comment (Grokzen) [inline setup.py:20]: If we have a requirements file, then this should be added there as well.

**Review by Grokzen — CHANGES_REQUESTED** [review]

File: setup.py (hunk 13 above)
Comment (holmboe) [reply setup.py:20]: Currently does not have a `requirements.txt`, out of scope for now.

File: holmscan/constants.py (hunk 5 above)
Comment (holmboe) [reply holmscan/constants.py:3]: I don't get the added benefit of this.

File: holmscan/controller.py (hunk 9 above)
Comment (holmboe) [reply holmscan/controller.py:36]: I don't get the added benefit of this either.

File: holmscan/scan.py (hunk 11 above)
Comment (holmboe) [reply holmscan/scan.py]: Will look into this. Stay tuned.

Comment (holmboe) [reply holmscan/scan.py]: I have now moved `.get()` and `.post()` into the controller object.

File: holmscan/controller.py (hunk 7 above)
Comment (holmboe) [reply holmscan/controller.py:29]: Leaving as is for now.

File: holmscan/controller.py (hunk 8 above)
Comment (holmboe) [reply holmscan/controller.py:32]: They are required configurations, or else the tool will ask for them. Leaving as is for now.

**Pull request merged**